manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java

```diff
@@ -0,0 +1,639 @@
+//            getLogger().warning(e.fi);
+        }
+        // Check if asset was found
+    }
+
+    /**
+     * Creates a new asset with the correct "hashed" Asset ID, its IMEI,
+     * in the realm the MQTT message of the device submitted,
+     * and the parsed list of attributes.
+     * //TODO: Add {@value DEVICE_SEND_COMMAND_ATTRIBUTE_NAME }, {@value DEVICE_RECEIVE_COMMAND_ATTRIBUTE_NAME} Attributes when creating Asset
+     *
+     * @param newDeviceId The ID of the device's Asset.
+     * @param newDeviceImei The IMEI of the device. If passed to
+     *                      {@link UniqueIdentifierGenerator#generateId(String)},
+     *                      it should always return {@code newDeviceId}.
+     * @param realm The realm to create the Asset in.
+     * @param attributes The attributes to insert in the Asset.
+     */
+    private void CreateNewAsset(String newDeviceId, String newDeviceImei, String realm, AttributeMap attributes) {
+        getLogger().info("Creating CarAsset with IMEI "+newDeviceImei);
+
+        CarAsset testAsset = new CarAsset("Teltonika Asset "+newDeviceImei)
+                .setRealm(realm)
+                .setId(newDeviceId);
+
+        //TODO: Maybe move these to getAttributes?
+        attributes.add(new Attribute<>("IMEI", ValueType.TEXT, newDeviceImei));
+        attributes.add(new Attribute<>("lastContact", ValueType.DATE_AND_TIME,
+                new Date(timerService.getCurrentTimeMillis())));
+
+        testAsset.addOrReplaceAttributes(attributes.stream().toArray(Attribute[]::new));
+        Asset<?> mergedAsset = assetStorageService.merge(testAsset);
+
+        if(mergedAsset != null){
+            getLogger().info("Created Asset through Teltonika: " + testAsset);
+//            getLogger().info()
+        }else{
+            getLogger().info("Failed to create Asset: " + testAsset);
+        }
+    }
+
+    /**
+     * Returns list of attributes depending on the Teltonika JSON Payload.
+     * Uses the logic and results from parsing the Teltonika Parameter IDs.
+     *
+     * @param payloadContent Payload coming from Teltonika device
+     * @return Map of {@link Attribute}s to be assigned to the {@link Asset}.
+     */
+    private AttributeMap getAttributesFromPayload(String payloadContent) throws JsonProcessingException {
+        HashMap<Integer, TeltonikaParameter> params = new HashMap<>();
+        ObjectMapper mapper = new ObjectMapper();
+        try {
+            // Parse file with Parameter details
+            TeltonikaParameter[] paramArray = mapper.readValue(getParameterFileString(), TeltonikaParameter[].class);
+
+            getLogger().info("Parsed "+paramArray.length+" Teltonika Parameters");
+
+            // Add each element to the HashMap, with the key being the unique parameter ID and the parameter
+            // being the value
+            for (TeltonikaParameter param : paramArray) {
+                params.put(param.getPropertyId(), param);
```

Comment:
There is probably an more efficient / elegant solution to this. Java streams?

Reply:
Implemented using Java streams